crates/fuel-core/src/graphql_api/database.rs

```diff
@@ -117,12 +119,56 @@ impl DatabaseBlocks for ReadView {
         height: Option<BlockHeight>,
         direction: IterDirection,
     ) -> BoxedIter<'_, StorageResult<CompressedBlock>> {
-        self.on_chain.blocks(height, direction)
+        // Chain together blocks from the off-chain db and the on-chain db
+        // The blocks in off-chain db, if any, are from time before regenesis
+
+        if let Some(height) = height {
+            match self.on_chain.first_height() {
+                Ok(onchain_start_height) => {
+                    match (height >= onchain_start_height, direction) {
+                        (true, IterDirection::Forward) => {
+                            self.on_chain.blocks(Some(height), direction)
+                        }
+                        (true, IterDirection::Reverse) => self
+                            .on_chain
+                            .blocks(Some(height), direction)
+                            .chain(self.off_chain.old_blocks(None, direction))
+                            .into_boxed(),
+                        (false, IterDirection::Forward) => self
+                            .off_chain
+                            .old_blocks(Some(height), direction)
+                            .chain(self.on_chain.blocks(None, direction))
+                            .into_boxed(),
+                        (false, IterDirection::Reverse) => {
+                            self.off_chain.old_blocks(Some(height), direction)
+                        }
+                    }
+                }
+                Err(err) => core::iter::once(Err(err)).into_boxed(),
+            }
+        } else {
+            match direction {
+                IterDirection::Forward => self
+                    .off_chain
+                    .old_blocks(height, direction)
+                    .chain(self.on_chain.blocks(height, direction))
+                    .into_boxed(),
+                IterDirection::Reverse => self
+                    .on_chain
+                    .blocks(height, direction)
+                    .chain(self.off_chain.old_blocks(height, direction))
+                    .into_boxed(),
+            }
+        }
     }
 
     fn latest_height(&self) -> StorageResult<BlockHeight> {
         self.on_chain.latest_height()
     }
+
+    fn first_height(&self) -> StorageResult<BlockHeight> {
```

Comment:
The `genesis_height` sounds better=) Plus, since we may have many genesis heights, maybe it would be nice to highlight it somehow `active_genesis_height`, `actual_genesis_height`, `on_chain_genesis_height`=)

Reply:
Maybe `latest_genesis_height`? I'll rename it.

Reply:
e7ef64c